mlflow/store/artifact/cloud_artifact_repo.py

```diff
@@ -0,0 +1,272 @@
+
+
+def _complete_futures(futures_dict, file):
+    """
+    Waits for the completion of all the futures in the given dictionary and returns
+    a tuple of two dictionaries. The first dictionary contains the results of the
+    futures (unordered) and the second contains the errors (unordered) that occurred
+    during the execution of the futures.
+    """
+    results = {}
+    errors = {}
+
+    with ArtifactProgressBar.chunks(
+        os.path.getsize(file),
+        f"Uploading {file}",
+        _MULTIPART_UPLOAD_CHUNK_SIZE,
+    ) as pbar:
+        for future in as_completed(futures_dict):
+            key = futures_dict[future]
+            try:
+                results[key] = future.result()
+                pbar.update()
+            except Exception as e:
+                errors[key] = repr(e)
+
+    return results, errors
+
+
+StagedArtifactUpload = namedtuple(
+    "StagedArtifactUpload",
+    [
+        # Local filesystem path of the source file to upload
+        "src_file_path",
+        # Base artifact URI-relative path specifying the upload destination
+        "artifact_file_path",
+    ],
+)
+
+
+class CloudArtifactRepository(ArtifactRepository):
+    def __init__(self, artifact_uri):
+        super().__init__(artifact_uri)
+        # Use an isolated thread pool executor for chunk uploads/downloads to avoid a deadlock
+        # caused by waiting for a chunk-upload/download task within a file-upload/download task.
+        # See https://superfastpython.com/threadpoolexecutor-deadlock/#Deadlock_1_Submit_and_Wait_for_a_Task_Within_a_Task
+        # for more details
+        self.chunk_thread_pool = self._create_thread_pool()
+
+    # Write APIs
+
+    def log_artifacts(self, local_dir, artifact_path=None):
+        """
+        Parallelized implementation of `log_artifacts`.
+        """
+
+        artifact_path = artifact_path or ""
+
+        staged_uploads = []
+        for dirpath, _, filenames in os.walk(local_dir):
+            artifact_subdir = artifact_path
+            if dirpath != local_dir:
+                rel_path = os.path.relpath(dirpath, local_dir)
+                rel_path = relative_path_to_artifact_path(rel_path)
+                artifact_subdir = posixpath.join(artifact_path, rel_path)
+            for name in filenames:
+                src_file_path = os.path.join(dirpath, name)
+                src_file_name = os.path.basename(src_file_path)
+                staged_uploads.append(
+                    StagedArtifactUpload(
+                        src_file_path=src_file_path,
+                        artifact_file_path=posixpath.join(artifact_subdir, src_file_name),
+                    )
+                )
+
+        # Join futures to ensure that all artifacts have been uploaded prior to returning
+        failed_uploads = {}
+
+        # For each batch of files, upload them in parallel and wait for completion
+        def upload_artifacts_iter():
```

Comment:
Not a blocker, but can we change this function to a method?

Reply:
I'll do this in an upcoming PR